### ISSUE TITLE
Expose Blocking/Non-blocking lock acquire semantics on RedLock context manager

### ISSUE THREAD

**Issue by sidmitra** [brainix/pottery#422]
**Is your feature request related to a problem? Please describe.**
I typically use(and advise others to use) the `with` statement for creating a redlock instance, to avoid forgetting to release it.  Using the with statement, you lose the ability to set lock acquiring semantics and timeout on the lock because you no longer call acquire() yourself but is automatically called in `Redlock.__enter__()`. So you are stuck with the default blocking indefinitely semantic.

**Describe the solution you'd like**

```python
from pottery import Redlock
from functools import partial

# Avoid passing around redis clients everywhere, by creating a partial
# IMHO helps keeps the lock interface smaller at the time of use, where a dev on the team
# doesn't need to worry about the redis client.
Lock = partial(Redlock, masters={redis})
NonBlockingLock = partial(Lock, blocking=False)
BlockingLock = partial(Lock, blocking=True, timeout=10)


# same as it works currently, with the defaults set
with Lock(key='printer'):
   critical_section()

with BlockingLock(key='printer'):
   critical_section()

with NonBlockingLock(key='printer'):
   another_critical_section()
```

**Describe alternatives you've considered**

- Currently i've written a wrapper on top with defaults typically used.

```
from contextlib import ContextDecorator
from pottery import Redlock

class Lock(ContextDecorator):
    def __init__(self, key: str, release_after_seconds, blocking=False, blocking_timeout_seconds = -1):
        self._lock = Redlock(
            key=self.key,
            masters={redis},
            auto_release_time=release_after_seconds * 1000,  # convert to milliseconds
        )
        self.blocking = blocking
        self.blocking_timeout_seconds = blocking_timeout_seconds

    def __enter__(self):
        if not self._lock.acquire(blocking=self.blocking, timeout=self.blocking_timeout_seconds):
            raise LockError
        return self

    def __exit__(self, *exc):
        self._lock.release()
        return False
```

- Probably a sub-class of Redlock that extends `__enter__` and `__exit__ ` would also work? But i saw a lot of name mangling happening, so wasn't sure i was going to mess something up. But will also try that solution once.

**Additional context**

This is more of a **nice-to-have**, since it's all possible to build on top of the Redlock primitive. I do think the context-manager is the proper way to use these features with block scoping, since in a larger dev team it's all too common for people to forget to call lock.release() or close file pointers etc. 

IMHO, it might also make the context-manager interface slightly more expressive and be able to modify behaviour using `functools.partial`. I'm all ears to see what you think, and issues you see.

PS:  If you see merit in this request, i can send a PR.

**Comment by brainix** [comment]
Hi, @sidmitra!

Thanks for the suggestion. I've tried to incorporate here: https://github.com/brainix/pottery/pull/424

Once I've merged that PR, you'll be able to instantiate Redlock like this:

```python
>>> redlock = Redlock(key='printer-lock', masters={redis}, context_manager_blocking=True, context_manager_timeout=0.2)
```

Then when you use that Redlock as a context manager, it'll use those blocking and timeout parameters.

Similarly, I've added the `blocking` and `timeout` arguments to the `synchronize()` decorator.

Does this satisfy your use case? How does the PR look?

Thanks again,
Raj

P.S. - I agree with your advice to use Redlock as a context manager to prevent bugs. I also love that you use `functools.partial()` to simplify Redlock instantiation and to reduce cognitive load.

**Comment by sidmitra** [comment]
@brainix Great. Will play around with it over the next day or so and leave a comment here. 

Appreciate the quick response, and excited to try it out in production. Currently we're doing rudimentary homegrown SETNX stuff, and i'm going to try pottery out next week.

**Comment by brainix** [comment]
Hi again, @sidmitra.

I've just released Pottery v1.3.0, which includes this new functionality.

Thanks for the suggestion,
Raj

**Comment by sidmitra** [comment]
Thanks for the release. Will update my project.
